tictactoeteam/Tic_Tac_Toe
# Database Connection, Player/Game Models and DAOs, Migration

Started the database system. The Game DAO needs a little more work, but I wanted to get these models to you guys ASAP so you can start using them.

## Overview
 - Added two new models: `Player` and `Game`
 - These models are saved to the `players` and `games` tables in postgres
 - These models can be used independently of the database in other parts of the program
 - To actually read/write/delete these models to/from the database, use `GameDao` and `PlayerDao`
 - Players and Games have IDs (UUIDs from pgcrypto)

## Players
 - Player passwords are hashed with Bcrypt. DO NOT MANIPULATE `hashedPassword` directly! Use `setPassword()` to set the user's password, and `checkPassword()` to check if a user-inputted password is correct. You do *not* have access to the user's password when you read from the database. You only have the ability to check if a user-inputted password matches the database

## Games
 - Games contain two foreign keys from `players`, for player X and player O, and also contain a byte[] of moves
 - `moves` is an array of the moves of a game. It is basically a log of each position (0-8) of where the players moved, in order. So an array of `{ 2, 5, 3}` indicates a game where X went top-right, O went middle-right, X went middle-left.

## Migrations
 - Wrote a very simple migration framework. To run a migration, start the server with `java -jar <jar> --server --migrate`.
 - The index of the last migration that was run is stored in postgres, so the server will be able to determine which migration to start with.
 - After this PR lands, DO NOT modify the Init script. The purpose of migrations is to provide an immutable, replayable history of database schemas
 - After this PR lands, if you need to modify the `Game` or `Player` schemas, you MUST create a new migration script with `ALTER TABLE` commands to update the schema

## Running the stack
 - We are using postgres for the database, this will 100% not work with any other database software
 - I highly recommend installing Docker and Docker Compose to quickly start the server and all backing services
 - Once docker-compose is installed, run `make build-docker` or `docker build .` to build the server, and `make up`, or `docker-compose up` to start the stack with all backing services

## Testing this PR
 - Build and run with `make build-docker` and `make up`
 - Make sure the migrations work fine, if you want to dive into `psql` to check the tables go ahead
 - If you want, try creating players, setting passwords, checking passwords, deleting/update players, etc.
